src/audio/volume/volume.c

```diff
@@ -1181,6 +1179,74 @@ static vol_zc_func vol_get_zc_function(struct comp_dev *dev)
 	return NULL;
 }
 
+/**
+ * \brief Set volume frames alignment limit.
+ * \param[in,out] source Source structure.
+ * \param[in,out] sink Sink structure.
+ */
+#if XCHAL_HAVE_HIFI3 || XCHAL_HAVE_HIFI4
+
+/**
+ * in HiFi3 or HiFi4, we use multi-way operation intrinsics, those operations
+ * require 8-byte aligned, and 5.1 format SSE audio asked for 16-byte aligned,
+ * so we should aligned with 8 or 16 bytes. Since there is at most
+ * SOF_IPC_MAX_CHANNELS channels, the following 2 arrays are the restriction
+ * tables of 16 and 32 bit per sample.
+ */
+
+const struct processing_alignment volume_align_16bit[SOF_IPC_MAX_CHANNELS] = {
+	/** to meet the requirement of x-byte aligned,the frame_size * frame_align
+	 * should be multiple of x
+	 */
+	/* channel_num, alignment_shift_idx */
+	{1, 2}, /*8-byte aligned, frame size: 2 bytes, frame_align: 4 frames*/
+	{2, 2}, /*16-byte aligned, frame size: 4 bytes, frame_align: 4 frames*/
+	{3, 2}, /*8-byte aligned, frame size: 6 bytes, frame_align: 4 frames*/
+	{4, 0}, /*8-byte aligned, frame size: 8 bytes, frame_align: 1 frames*/
+	{5, 2}, /*8-byte aligned, frame size: 10 bytes, frame_align: 4 frames*/
+	{6, 2}, /*16-byte aligned, frame size: 12 bytes, frame_align: 4 frames*/
+	{7, 2}, /*8-byte aligned, frame size: 14 bytes, frame_align: 4 frames*/
+	{8, 0} /*8-byte aligned, frame size: 16 bytes, frame_align: 1 frames*/
+};
+
+const struct processing_alignment volume_align_32bit[SOF_IPC_MAX_CHANNELS] = {
+	{1, 1}, /*8-byte aligned, frame size: 4 bytes, frame_align: 2 frames*/
+	{2, 1}, /*16-byte aligned, frame size: 8 bytes, frame_align: 2 frames*/
+	{3, 1}, /*8-byte aligned, frame size: 12 bytes, frame_align: 2 frames*/
+	{4, 0}, /*8-byte aligned, frame size: 16 bytes, frame_align: 1 frames*/
+	{5, 1}, /*8-byte aligned, frame size: 20 bytes, frame_align: 2 frames*/
+	{6, 1}, /*16-byte aligned, frame size: 24 bytes, frame_align: 2 frames*/
+	{7, 1}, /*8-byte aligned, frame size: 28 bytes, frame_align: 2 frames*/
+	{8, 0} /*8-byte aligned, frame size: 32 bytes, frame_align: 1 frames*/
+};
+
+static void volume_get_alignment(struct audio_stream *source,
+				 struct audio_stream *sink)
+{
+	uint16_t shift_idx = 0;
+	struct processing_alignment *volume_align;
+	uint16_t sample_size = get_sample_bytes(source->frame_fmt);
+
+	if (sample_size == 2)
+		volume_align = (struct processing_alignment *)volume_align_16bit;
+	else
+		volume_align = (struct processing_alignment *)volume_align_32bit;
+
+	shift_idx = volume_align[source->channels].alignment_shift_idx;
```

Comment:
its best to not assume the ordering of the arrays and search the array for the matching entry. This ways its secure to channels < min or > max.